### PR TITLE
Use correct string length and number format

### LINKED ISSUE #3810: Unify string representation of floats in userdata objects (vector3 etc) (DEF-1169)
Consider this:

local a = vmath.vector3(0, 0, math.pi)
pprint(a)

DEBUG:SCRIPT: vmath.vector3(0, 0, 3.1415927410126)

And the following:

print("vector is: " .. a)

DEBUG:SCRIPT: vector is: [0.000000, 0.000000, 3.141593]

Notice that the float numbers are printed differently. It would be nice if they were given the same representation in all cases.

### PR DESCRIPTION
**Pull request by britzl**
Fixes #3810

### CODE REVIEW

File: engine/script/src/script_vmath.cpp [hunk 1]
```diff
@@ -300,8 +300,9 @@ namespace dmScript
         size_t size = 0;
         const char* s = luaL_checklstring(L, 1, &size);
         Vectormath::Aos::Vector3* v = CheckVector3(L, 2);
-        char* buffer = new char[size + 61];
-        dmSnPrintf(buffer, size, "%svmath.%s(%f, %f, %f)", s, SCRIPT_TYPE_NAME_VECTOR3, v->getX(), v->getY(), v->getZ());
+        char* buffer = new char[size + 68];
+        // Use same format as Lua when converting number to string (from LUA_NUMBER_FMT in luaconf.h)
+        dmSnPrintf(buffer, size + 68, "%svmath.%s(%.14g, %.14g, %.14g)", s, SCRIPT_TYPE_NAME_VECTOR3, v->getX(), v->getY(), v->getZ());
```
Comment (JCash) [inline engine/script/src/script_vmath.cpp:305]: Fyi, 32 bit floating point numbers doesn't have 14 bits of precision.

Also, what is 68 based off of? It increased by 7? 7 is not a multiple of 3, so I'm curious.

Like Lua, I'd recommend constants (e.g. at the top of the file), where we can also document them

File: engine/script/src/test/test_matrix4.lua [hunk 2]
```diff
@@ -156,3 +156,14 @@ m = vmath.matrix4_translation(vmath.vector4(8,7,6,-1))
...
+-- tostring and concat
+m = vmath.matrix4_translation(vmath.vector4(8,7,6,-1))
+assert(("foo " .. tostring(m)) == "foo vmath.matrix4(1, 0, 0, 8, 0, 1, 0, 7, 0, 0, 1, 6, 0, 0, 0, 1)")
+assert(("foo " .. m) == "foo vmath.matrix4(1, 0, 0, 8, 0, 1, 0, 7, 0, 0, 1, 6, 0, 0, 0, 1)")
+m = vmath.matrix4()
+m.c0 = vmath.vector4(-10.01,-10.01,-10.01,-10.01)
+m.c1 = vmath.vector4(-10.01,-10.01,-10.01,-10.01)
+m.c2 = vmath.vector4(-10.01,-10.01,-10.01,-10.01)
+m.c3 = vmath.vector4(-10.01,-10.01,-10.01,-10.01)
+assert(tostring(m) == ("" .. m))
```
Comment (JCash) [inline engine/script/src/test/test_matrix4.lua:169]: 👍

**Review by JCash — DISMISSED** [review]

File: engine/script/src/script_vmath.cpp (hunk 1 above)
Comment (britzl) [reply engine/script/src/script_vmath.cpp:305]: > 32 bit floating point numbers doesn't have 14 bits of precision

The %.14g format comes from here:

https://github.com/defold/defold/blob/dev/engine/lua/src/dmsdk/lua/luaconf.h#L522.

We use lua_pushfstring() for Vector3_tostring() so I thought it best to use the same format specifier for our own conversion.

NOTE: We call `lua_pushfstring` with "%f" which is Lua's own format specifier which under the hood translates to "%.14g".

> Also, what is 68 based off of?

Worst case scenario when converted to String is (if I'm not mistaken) something like this:

`vmath.vector3(-10.010000228882, -10.010000228882, -10.010000228882)`

That is 67 characters, and then we need another to have room for the terminating null character.

Comment (JCash) [reply engine/script/src/script_vmath.cpp:305]: You specify 14 decimals, but in your example there is only 12?

My main point is that there should be a constant defining how many characters a float can take.
E..g "MAX_CHARS_PER_FLOAT = x"
and
MAX_VEXTOR3_STRING = MAX_CHARS_PER_FLOAT * 3 + ...

Comment (britzl) [reply engine/script/src/script_vmath.cpp:305]: Hmm, you're right.

File: engine/script/src/script_vmath.cpp [hunk 3]
```diff
@@ -71,6 +71,17 @@ namespace dmScript
...
+#define STRING_FORMAT_CONCAT_VECTOR3 "%svmath.vector3(%.14g, %.14g, %.14g)"
+#define STRING_FORMAT_CONCAT_VECTOR4 "%svmath.vector4(%.14g, %.14g, %.14g, %.14g)"
+#define STRING_FORMAT_CONCAT_QUAT    "%svmath.quat(%.14g, %.14g, %.14g, %.14g)"
+#define STRING_FORMAT_CONCAT_MATRIX4 "%svmath.matrix4(%.14g, %.14g, %.14g, %.14g, %.14g, %.14g, %.14g, %.14g, %.14g, %.14g, %.14g, %.14g, %.14g, %.14g, %.14g, %.14g)"
+
+#define MAX_CHARS_PER_FLOAT   19 // 1.4012984643248e-45
+#define MAX_CHARS_PER_VECTOR3 13 + 1 + (MAX_CHARS_PER_FLOAT * 3) + (2 * 2) + 1   // vmath.vector3(x, y, z)
+#define MAX_CHARS_PER_VECTOR4 13 + 1 + (MAX_CHARS_PER_FLOAT * 4) + (2 * 3) + 1   // vmath.vector4(x, y, z, w)
+#define MAX_CHARS_PER_QUAT    10 + 1 + (MAX_CHARS_PER_FLOAT * 4) + (2 * 3) + 1   // vmath.quat(x, y, z, w)
+#define MAX_CHARS_PER_MATRIX4 13 + 1 + (MAX_CHARS_PER_FLOAT * 16) + (2 * 15) + 1 // vmath.matrix4(m00, m01, ..., m33)
```
Comment (JCash) [inline engine/script/src/script_vmath.cpp:83]: 👍

**Review by JCash — APPROVED** [review]
Looks good!
Thanks!

**Pull request merged**